### PR TITLE
Use pagination on history tab

### LINKED ISSUE #44993: Add pagination support for FetchCommits API call in the History tab
# Problem

Currently we allow users to set a custom (1 to 100) page size for loading commits at a time in the history tab, but each subsequent request to fetch the next N commits still fetches 0 to N commits which makes each subsequent request linearly more and more expensive.

While the setting to customise a smaller page size does help the first few times a user is exploring the history tab, it does not solve the core problem.

# Proposal

Implement pagination so that subsequent requests only fetch the next N slice.

Related to #42370 and may help closing it as well

### PR DESCRIPTION
**Pull request by indradhanush**
History-pagination
Fixes #44993.

Follow up to #45157.

## What

Use `afterCursor` to skip the first N commits for subsequent requests

## Why

- We don't need to fetch the same commits for subsequent requests
- We only fetch `history.defaultPageSize` number of commits each time (defaults to 100 is not set) instead of doubling the number of commits for subsequent requests -> consistent behaviour -> win win

## How

Replace the use of `FileredConnection` with `useShowMorePagination` which can make use of the `endCursor` from `PageInfo`.

## Extra benefits

Now responses are cached, so if the user closes the history tab and opens it again, they will see the commits without the need of a new network call (current behavior before this PR)

## Test plan

Tested locally.

## App preview:

- [Web](https://sg-web-ig-frontend-history-pagination.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| -0.00% (-0.08 kb) | 0.00% (+0.03 kb) | 0.00% (+0.11 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/ig/frontend-history-pagination/compare-report.html;tab=live_object) for a full comparison between the commits f30a55f and aaceb90 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/web/src/repo/RepoRevisionSidebarCommits.tsx [hunk 1]
```diff
@@ -67,83 +67,94 @@ interface Props extends Partial<RevisionSpec>, FileSpec {
...
-        />
+        <ConnectionContainer>
+            {error && <ErrorAlert error={error} />}
+            {connection &&
+                connection.nodes.map(node => (
+                    <CommitNode
+                        node={node}
+                        location={props.location}
+                        preferAbsoluteTimestamps={props.preferAbsoluteTimestamps}
+                    />
```
Comment (philipp-spiess) [inline client/web/src/repo/RepoRevisionSidebarCommits.tsx:118]: The CommitNode must have `key` so React knows how to diff this properly. Not sure if this field already has an ID but if not use _anything_ that is unique within this list. It does not need to be globally unique. 

```suggestion
                connection.nodes.map(node => (
                    <CommitNode
                        key={node.id}
                        node={node}
                        location={props.location}
                        preferAbsoluteTimestamps={props.preferAbsoluteTimestamps}
                    />
```

File: client/web/src/components/FilteredConnection/hooks/useShowMorePagination.ts [hunk 2]
```diff
@@ -41,6 +41,12 @@ interface UseShowMorePaginationConfig<TResult> {
     pollInterval?: number
     /** Allows running an optional callback on any successful request */
     onCompleted?: (data: TResult) => void
+
+    // customAfterCursor is used to indicate that a custom field instead of the
+    // standard "after" field is used to for pagination. This is typically a
+    // workaround for existing APIs where after may already be in use for
+    // another field.
+    customAfterCursor?: boolean
```
Comment (philipp-spiess) [inline client/web/src/components/FilteredConnection/hooks/useShowMorePagination.ts:49]: I think the name is not super clear now with the boolean option, do you have another idea? Maybe `useAlternativeAfterCursor`? 😅

**Review by philipp-spiess — APPROVED** [review]
Great cleanup! Thank you for the work :)

**Review by cbart — COMMENTED** [review]
Looks good to me! Logic is pretty clear.

I do not know much about testing React. Do you think this logic is complex enough to warrant unit-testing, (if those are available)?

**Comment by indradhanush** [comment]
> I do not know much about testing React. Do you think this logic is complex enough to warrant unit-testing, (if those are available)?

Unit tests are always good I suppose. But I'm quite a React noob myself. The logic is simpler IMO and we will soon (a couple of releases after the upcoming one) be able to fix the root cause for the hack.

**Pull request merged**